Sources/AsyncHTTPClient/RequestBag.swift

```diff
@@ -0,0 +1,843 @@
+            func executeOnEL() {
+                do {
+                    let response = try self.delegate.didFinishRequest(task: task)
+                    self.task.promise.succeed(response)
+                } catch {
+                    self.task.promise.fail(error)
+                }
+            }
+
+            if self.task.eventLoop.inEventLoop {
+                executeOnEL()
+            } else {
+                self.task.eventLoop.execute {
+                    executeOnEL()
+                }
+            }
+        case .failTask(let error):
+            self.failTask(error)
+        case .requestMoreFromExecutor(let executor):
+            executor.demandResponseBodyStream(self)
+        }
+    }
+
+    func fail(_ error: Error) {
+        let (queuer, executor, forward) = self.stateLock.withLock {
+            () -> (HTTPRequestScheduler?, HTTPRequestExecutor?, Bool) in
+
+            switch self._state {
+            case .initialized:
+                self._state = .finished(error: error)
+                return (nil, nil, true)
+            case .queued(let queuer):
+                self._state = .finished(error: error)
+                return (queuer, nil, true)
+            case .executing(let executor, let requestState, .buffering(_, next: .eof)):
+                self._state = .executing(executor, requestState, .buffering(.init(), next: .error(error)))
+                return (nil, executor, false)
+            case .executing(let executor, let requestState, .buffering(_, next: .askExecutorForMore)):
+                self._state = .executing(executor, requestState, .buffering(.init(), next: .error(error)))
+                return (nil, executor, false)
+            case .executing(let executor, _, .buffering(_, next: .error(_))):
+                // this would override another error, let's keep the first one
+                return (nil, executor, false)
+
+            case .executing(let executor, _, .initialized):
+                self._state = .finished(error: error)
+                return (nil, executor, true)
+
+            case .executing(let executor, _, .waitingForRemote(_)):
+                self._state = .finished(error: error)
+                return (nil, executor, true)
+
+            case .redirected:
+                self._state = .finished(error: error)
+                return (nil, nil, true)
+
+            case .finished(.none):
+                // An error occured after the request has finished. Ignore...
+                return (nil, nil, false)
+
+            case .finished(.some(_)):
+                // this might happen, if the stream consumer has failed... let's just drop the data
+                return (nil, nil, false)
+
+            case .modifying:
+                preconditionFailure("Invalid state")
+            }
+        }
+
+        queuer?.cancelRequest(self)
+        executor?.cancelRequest(self)
+
+        if forward {
+            self.failTask(error)
+        }
+    }
+}
```

Comment:
I think this object would get _substantially_ cleaner if the request/response mini-state machines were codified and made explicit. This would shift much of the functionality out, which would make our lives a good deal easier.